hbase_storage/src/main/java/com/splicemachine/si/impl/region/SamplingFilter.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2012 - 2019 Splice Machine, Inc.
+ *
+ * This file is part of Splice Machine.
+ * Splice Machine is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3, or (at your option) any later version.
+ * Splice Machine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.splicemachine.si.impl.region;
+
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.exceptions.DeserializationException;
+import org.apache.hadoop.hbase.filter.FilterBase;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.spark.util.random.BernoulliSampler;
+import org.apache.spark.util.random.RandomSampler;
+import scala.reflect.ClassTag;
+
+import java.io.IOException;
+import java.io.Serializable;
+
+public class SamplingFilter extends FilterBase implements Serializable {
+
+    private double rate;
+    private RandomSampler sampler;
+
+    public SamplingFilter(){
+    }
+
+    public SamplingFilter(double rate) {
+        this.rate = rate;
+        ClassTag<String> tag = scala.reflect.ClassTag$.MODULE$.apply(String.class);
+        this.sampler = new BernoulliSampler(rate, tag);
+    }
+
+    @Override
+    public boolean filterRowKey(byte[] buffer, int offset, int length) throws IOException {
+        if (length == 0) return false; // make sure we don't remove special Memstore BEGIN / FLUSH rows
```

Comment:
Is this sufficient to retain the special rows?  For example, a special FLUSH row could also be built with a non-zero rowkey:
```
            if (flushAlreadyReturned) {
                return new KeyValue(Bytes.toBytes(counter),ClientRegionConstants.FLUSH,ClientRegionConstants.FLUSH, 0l,ClientRegionConstants.FLUSH);
```
If the filtering were moved to filterKeyValue, then the Cell could be checked to see if it contains the special FLUSH column family, and avoid filtering it.

Reply:
Sorry, I missed this comment. I'd rather keep using filterRowKey because it has lower overhead and we can filter out rows without having to reconstruct the column families and all that. Those other rows that could get generated will eventually make it through (we increment the counter and regenerate them until the scanner is closed), the only case where this could be bad is if the selectivity is very very low, it might take a while until one of those rows gets picked up. However very, very low selectivities aren't much faster than low selectivities, so I don't expect this to be a big problem.

Reply:
OK, thanks for the explanation. That sounds good to me.